### ISSUE TITLE
[FEA] Refactor cuSignal Python code - CUDA Kernel Organization

### ISSUE THREAD

**Issue by awthomp** [rapidsai/cusignal#63]
cuSignal was originally created as a direct port of SciPy Signal, including leveraging their [software library design decisions](https://github.com/scipy/scipy/tree/master/scipy/signal). With the growth of cuSignal and introduction of more raw CuPy CUDA kernels, the cuSignal codebase is growing difficult to maintain and document. Further, many seemingly disparate functions are housed in meta-files like `signaltools.py`. For example, `fftconvolve` and `resample` - the first of which is a convolution function and the latter a filtering operation, can both be found in `signaltools.py`.

Targeting cuSignal v0.15, I'd like to propose a code re-organization and am open to suggestions on how best to accomplish this. Mirroring [cuDF](https://github.com/rapidsai/cudf/tree/branch-0.14/python/cudf/cudf), I'd like to have a folder called `_cuda` that consists of our Raw CUDA kernels. Further, mirroring packages like [numpy](https://github.com/numpy/numpy/tree/master/numpy), I'd like to create folders within `cusignal/python/cusignal` that organize by topic. Within each folder, the topic should be further subdivided into related groups, as needed.

Examining the SciPy Signal [documentation](https://docs.scipy.org/doc/scipy/reference/signal.html) exposes the following topics currently relevant to cuSignal:
1. convolution
2. b-splines
3. filtering
4. filter design
5. waveforms
6. window functions
7. wavelets
8. peak finding
9. spectral analysis

We've also included an additional `acoustics` module and pytests functionality in cuSignal 0.13. cuSignal 0.14 will bring pytest benchmarks. Additional folders suggested for the refactor:
* arrays (for cusignal `arraytools`)
* acoustics
* benchmark
* test
* _cuda (for raw CuPy CUDA kernels, mentioned before)

Within each of these 'topic folders', we need to further group. For example, convolution contains
* `convolve`
* `fftconvolve`
* `convolve2d`
* `correlate`
* `correlate2d`

among others. In this case, I feel it's logical to divide these functions into the following groups:
1. `convolve.py`: `convolve`, `fftconvolve`, `convolve2d`
2. `correlate.py`: `correlate`, `correlate2d`

To be explicit, this will create the following directory structure:
`cusignal/python/cusignal/convolution` with `convolve.py` and `correlate.py` housed in this folder.

I'll take an initial swing at the organization and would appreciate input from the user community. With this resultant PR, we can address other issues, like https://github.com/rapidsai/cusignal/issues/58 and https://github.com/rapidsai/cusignal/issues/54

CC: @mnicely @dicta @boonedoggle @BradReesWork

**Comment by awthomp** [comment]
We are halfway there! With we reorganized the directory structure and improved documentation. 

The bulk of cuSignal codebase has now been reorganized, but the Numba CUDA and raw CuPy CUDA kernels still reside in `_upfirdn.py`, `_spectral.py`, and `_signaltools.py`. We still need to reorganize:

- [ ] `convolve2d`
- [ ] `convolve1d`
- [ ] `correlate2d`
- [ ] `correlate1d`
- [ ] `upfirdn`
- [ ] `lombscargle`

Spoke to @dicta this morning, and he suggested keeping the CUDA kernels as close to the given operation as needed. In other words, if you're working on convolve2d, it would make more sense for the CUDA kernel to be in the convolution rather than cuda.

Concerning kernel compilation and caching, it would make sense to have a newly defined class to manage these operations and management. Here, it'd make sense to store this in a cuda folder and potentially move some of the memory management stuff (currently in utils.arraytools) there.

In summary, I think the current plan is:

* Keep CUDA kernels and wrappers close to the given operation. E.g. create a file called convolve_cuda.py in convolution for 1D/2D convolution and correlate_cuda.py in convolution for 1D/2D correlation.
* Create cuda folder and move functions like populate_kernel_cache and precompile_kernels to a newly defined class.
